### PR TITLE
GG-33033 Partition Reconciliation should restore internal caches

### PR DESCRIPTION
**Pull request by ashapkin**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java [hunk 1]
```diff
@@ -288,14 +290,18 @@ public RepairJob(Map<VersionedKey, Map<UUID, VersionedValue>> data, String cache
                                 cacheObjCtx,
                                 ownersNodesSize);
 
-                            keyWasSuccessfullyFixed = ignite.cache(cacheName).withKeepBinary().<RepairEntryProcessor.RepairStatus>invoke(
+                            EntryProcessorResult<Object> invoke = ctx.cache().keepBinary().invoke(
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java:293]: Why not to use ignite.cachex(cacheName) instead of ctx.cache()?
I'd also consider adding withKeepBinary to IgniteInternalCache in order to ease testing, issue related to mocking, cause keepBinary() itself in final.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/checker/tasks/PartitionReconciliationProcessorTask.java [hunk 2]
```diff
@@ -19,13 +19,15 @@
...
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.UUID;
+
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/checker/tasks/PartitionReconciliationProcessorTask.java:30]: Please remove unnecessary blank line.

File: modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java [hunk 3]
```diff
@@ -74,6 +76,8 @@
         GridTestUtils.addTestIfNeeded(suite, ReconciliationResourceLimitedJobTest.class, ignoredTests);
         GridTestUtils.addTestIfNeeded(suite, PartitionReconciliationFastCheckTest.class, ignoredTests);
         GridTestUtils.addTestIfNeeded(suite, PartitionReconciliationCompactCollectorTest.class, ignoredTests);
+        GridTestUtils.addTestIfNeeded(suite, PartitionReconciliationAtomicLongDataStructureTest.class, ignoredTests);
+        GridTestUtils.addTestIfNeeded(suite, PartitionReconciliationSetDataStructureTest.class, ignoredTests);
```
Comment (sanpwc) [inline modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java:80]: Please consider adding sys caches to partition reconciliation stress tests.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerPartitionReconciliationExtendedTest.java [hunk 4]
```diff
@@ -216,6 +258,31 @@ public void testWorkWithAllSetOfCachesIfParameterAbsent() throws Exception {
         assertEquals(usedCaches.size(), setOfCaches.size());
     }
 
+    /**
+     * Creates internal and custom caches and runs reconciliation cmd with args.
+     */
+    private Set<String> getUsedCachesForArgs(String... args) throws Exception {
+        Set<String> usedCaches = new HashSet<>();
+        LogListener lsnr = fillCacheNames(usedCaches);
+        log.registerListener(lsnr);
```
Comment (slukyano) [inline modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerPartitionReconciliationExtendedTest.java:267]: I think you want to unregister the log listener at the end of this method

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
```
Comment (slukyano) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java:83]: I think you're supposed to do this only in afterTest, and just trust other tests to have shut down themselves.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
```
Comment (slukyano) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java:90]: I think nowadays you're supposed to use `@After` instead of overriding afterTest. Also, if you're overriding you need to call `super.afterTest()` at the end.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        GridCacheContext[] nodeCacheCtxs = new GridCacheContext[NODES_CNT];
+
+        for (int i = 0; i < NODES_CNT; i++)
+            nodeCacheCtxs[i] = grid(i).cachex(cacheName).context();
+
+        for (int i = firstBrokenKey; i < firstBrokenKey + BROKEN_KEYS_CNT; i++) {
+            Class<?> aCls = Class.forName(GridCacheInternalKeyImpl.class.getCanonicalName());
+            for (Constructor<?> constructor : aCls.getDeclaredConstructors()) {
+                constructor.setAccessible(true);
+                if (constructor.getParameterCount() == 2) {
```
Comment (slukyano) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java:137]: Better to find the constructor before the loop instead of looking for it for each key. Also, please add a comment with the signature you expect.

Also, why is this needed? I see both GridCacheInternalKeyImpl constructors are public

**Review by slukyano — CHANGES_REQUESTED** [review]
Please add a test for --fast-check with sys caches.

File: modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerPartitionReconciliationExtendedTest.java (hunk 4 above)
Comment (ashapkin) [reply modules/control-utility/src/test/java/org/apache/ignite/util/GridCommandHandlerPartitionReconciliationExtendedTest.java:267]: Fixed

File: modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java (hunk 3 above)
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java:82]: Added, but --repair seems to be failing on AtomicLong, investigating

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java (hunk 1 above)
Comment (ashapkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java]: Done, cachex#keepBinary is not final, I've managed to mock it, I think there is no need in adding #withKeepBinary to internalCache anymore.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java (hunk 5 above)
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java]: I decided to keep the test in sync with other onces. Probably we can refactor all of them in a package separately.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java (hunk 6 above)
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAtomicLongDataStructureTest.java]: I decided to keep the test in sync with other ones. Probably we can refactor all of them in a package separately.

File: modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java (hunk 3 above)
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/testsuites/PartitionReconciliationTestSuite.java:82]: I'll create a separate ticket for that. For now, it seems to be working in case of a non-stress scenario.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/checker/tasks/PartitionReconciliationProcessorTask.java [hunk 8]
```diff
@@ -212,13 +216,17 @@ public PartitionReconciliationJob(
         @Override public T2<String, ExecutionResult<ReconciliationAffectedEntries>> execute() throws IgniteException {
             Set<String> caches = new HashSet<>();
 
+            Collection<String> cacheNames = F.viewReadOnly(ignite.context().cache().cacheDescriptors().values(),
```
Comment (sk0x50) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/checker/tasks/PartitionReconciliationProcessorTask.java:219]: Perhaps, you can use `ignite.context().cache().cacheNames();` instead of this.

Comment (ashapkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/verify/checker/tasks/PartitionReconciliationProcessorTask.java]: Done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java [hunk 9]
```diff
@@ -49,6 +49,7 @@
 import org.apache.ignite.internal.util.typedef.internal.U;
 import org.apache.ignite.resources.IgniteInstanceResource;
 import org.apache.ignite.resources.LoggerResource;
+import javax.cache.processor.EntryProcessorResult;
```
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTask.java]: Done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTaskTest.java [hunk 10]
```diff
@@ -301,18 +301,20 @@ private VersionedValue versionedValue(String val, int ver) {
...
-        when(cacheMock.invoke(any(), any(EntryProcessor.class))).thenReturn(invokeReturnFixed ?
-            RepairEntryProcessor.RepairStatus.SUCCESS : RepairEntryProcessor.RepairStatus.FAIL);
+        when(internalCacheMock.keepBinary()).thenReturn(internalCacheMock);
+        try {
+            when(internalCacheMock.invoke(any(), any(EntryProcessor.class))).thenReturn(invokeReturnFixed
+                ? CacheInvokeResult.fromResult(RepairEntryProcessor.RepairStatus.SUCCESS)
+                : CacheInvokeResult.fromResult(RepairEntryProcessor.RepairStatus.FAIL));
+        }
+        catch (IgniteCheckedException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/tasks/RepairRequestTaskTest.java]: Thx

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAbstractTest.java [hunk 11]
```diff
@@ -119,10 +133,24 @@ public static void assertResultContainsConflictKeys(
         String cacheName,
         Set<Integer> keys
     ) {
+        Pattern p = Pattern.compile("name=(\\d+)");
```
Comment (sk0x50) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAbstractTest.java:136]: Please remove unused variable

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationSetDataStructureTest.java [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+import org.junit.runners.Parameterized;
+
+import static org.apache.ignite.IgniteSystemProperties.IGNITE_SENSITIVE_DATA_LOGGING;
+
+/**
+ * Tests that reconciliation works with data collections.
+ */
+@RunWith(Parameterized.class)
+public class PartitionReconciliationSetDataStructureTest extends PartitionReconciliationAbstractTest {
+    /**
```
Comment (sk0x50) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationSetDataStructureTest.java:50]: Please use one-line comments here and below.

Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationSetDataStructureTest.java]: Done

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAbstractTest.java (hunk 11 above)
Comment (ashapkin) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/checker/processor/PartitionReconciliationAbstractTest.java]: Thx
